### DISCUSSION TITLE
How to get the html string?

### DISCUSSION THREAD

**Discussion by langsz** [Milkdown/milkdown#466] [Q&A] [answered]
Is there any api in mildown to get the html string?

**Comment by Saul-Mirone** [comment]
Why not just use `document.querySelector('.milkdown').outerHTML`?

**Comment by Saul-Mirone** [comment, marked as answer]
You can also write a simple action which will be provided officially in v6.0:

```typescript
/* Copyright 2021, Milkdown by Mirone. */
import { Ctx, editorViewCtx, schemaCtx } from '@milkdown/core';
import { DOMSerializer } from '@milkdown/prose';

export const getHTML = () => (ctx: Ctx) => {
    const div = document.createElement('div');
    const schema = ctx.get(schemaCtx);
    const view = ctx.get(editorViewCtx);
    const fragment = DOMSerializer.fromSchema(schema).serializeFragment(view.state.doc.content);

    div.appendChild(fragment);

    return div.innerHTML;
};

const html = editor.action(getHTML());
```

**Reply by langsz** [reply]
Get it

**Reply by seeARMS** [reply]
Is this now officially provided in the v6 release? I couldn't find any method involving `getHTML`. 

EDIT: Nevermind: https://github.com/Saul-Mirone/milkdown/blob/5ab806086322e23f851749c2c10b8976e165b896/packages/utils/src/macro/get-html.ts

**Reply by langsz** [reply]
> You can also write a simple action which will be provided officially in v6.0:
> 
> ```ts
> /* Copyright 2021, Milkdown by Mirone. */
> import { Ctx, editorViewCtx, schemaCtx } from '@milkdown/core';
> import { DOMSerializer } from '@milkdown/prose';
> 
> export const getHTML = () => (ctx: Ctx) => {
>     const div = document.createElement('div');
>     const schema = ctx.get(schemaCtx);
>     const view = ctx.get(editorViewCtx);
>     const fragment = DOMSerializer.fromSchema(schema).serializeFragment(view.state.doc.content);
> 
>     div.appendChild(fragment);
> 
>     return div.innerHTML;
> };
> 
> const html = editor.action(getHTML());
> ```

@Saul-Mirone I use the code to get the html, but there is some different with the html  in the editor.
for example,
![image](https://user-images.githubusercontent.com/31126835/167987159-0f45feec-64cc-4b26-a5eb-235142486c3e.png)
when checked the html elements,
![image](https://user-images.githubusercontent.com/31126835/167987273-6dd2744b-80da-4635-aeae-50e4af80c82b.png)

Do you have any method make them in the same?

**Reply by Saul-Mirone** [reply]
Because some of the plugins are rendered by [nodeviews in prosemirror](https://prosemirror.net/docs/ref/#view.EditorProps.nodeViews). And when DOM seriailizer parse them, they'll rendered by toDOM instead of nodeViews. I'm still thinking a good solution about this.

**Reply by langsz** [reply]
> Because some of the plugins are rendered by [nodeviews in prosemirror](https://prosemirror.net/docs/ref/#view.EditorProps.nodeViews). And when DOM seriailizer parse them, they'll rendered by toDOM instead of nodeViews. I'm still thinking a good solution about this.

Looking forware to your good solution.Thank you

**Reply by invokerzaa** [reply]
Just wanted to kindly check if there's any progress on this? I'm still thinking about a good solution and would love to hear your thoughts.
